File: ch06/41_1/FindNumbersWithSum.py
#####题目#####
# 输入一个递增排序的数组和一个数字S，在数组中查找两个数，是的他们的和正好是S，如果有多对数字的和等于S，输出两个数的乘积最小的。
# 提示：乘积最小的其实就是最开始遇到的
# -*- coding:utf-8 -*-
class Solution:
    def FindNumbersWithSum(self, array, tsum):
        # write code here
        re = []
        arrLen = len(array)
        if arrLen == 0 or tsum <= array[0]:
            return re

        start = 0
        end = arrLen - 1
        while start < end:
            curSum = array[start] + array[end]
            if  curSum == tsum:
                re = [array[start], array[end]]
                break
            elif curSum < tsum:
                start += 1
            else:
                end -= 1
        return re

File: ch06/41_1/FindNumbersWithSum.cpp
class Solution {
public:
    vector<int> FindNumbersWithSum(vector<int> array,int sum) {
        ///////a+b = sum, 那么有个结论就是a和b相差越远，成绩越小，刚好就是夹逼法的第一对
        vector<int> reVec;//用于返回的向量
       
        int len = array.size();//返回元素的个数
        if(len<2)
            return reVec;
        
        int begin = 0;
        int end = len - 1;
        while(begin < end){
            if(array[begin] + array[end] == sum){
                reVec.push_back(array[begin]);
                reVec.push_back(array[end]);
                break;
               // return reVec;
            }
            
            while(array[begin] + array[end] < sum){
                begin ++;
            }
            
            while(array[begin] + array[end] > sum){
                end --;                
            }
        }
        
        return reVec;
    }
};